### PR TITLE
Release 0.4.1

### PR DESCRIPTION
**Pull request by GearsAD**
# Release Notes
- Can get summary and summary graph from a DFG object
- DFG plotting functions
- Fixed issue with exports for DFGFactor
- Deprecated MetaGraphsDFG and SymbolDFG
- Factor and variable packing provided for users to easily pack and unpack DFG elements
- Improved tests and greater coverage of functionality
- Lot of function documentation

### CODE REVIEW

File: src/CloudGraphsDFG/services/CloudGraphsDFG.jl [hunk 1]
```diff
@@ -801,6 +787,33 @@ function getAdjacencyMatrix(dfg::CloudGraphsDFG)::Matrix{Union{Nothing, Symbol}}
...
+        error(string(nodes.errors))
+    end
+    # Add in the relationships
+    varRel = Symbol.(map(node -> node["row"][1], nodes.results[1]["data"]))
+    factRel = Symbol.(map(node -> node["row"][2], nodes.results[1]["data"]))
+    for i = 1:length(varRel)
+        adjMat[fDict[factRel[i]], vDict[varRel[i]]] = 1
+    end
+
+    return adjMat, varLabels, factLabels
```
Comment (dehann) [inline src/CloudGraphsDFG/services/CloudGraphsDFG.jl]: IIF still using Dense,  that can easily be done with constructors when we get there:
```julia
Ad = Matrix(sparse)
```

File: src/Common.jl [hunk 2]
```diff
@@ -88,6 +89,24 @@ function sortVarNested(vars::Vector{Symbol})::Vector{Symbol}
...
+
+Example
+
+`sortDFG(ls(dfg))`
+
+Related
+
+ls, lsf
+"""
+sortDFG(vars::Vector{Symbol})::Vector{Symbol} = sortVarNested(vars)
```
Comment (dehann) [inline src/Common.jl]: finally, awesome thanks!

File: src/FileDFG/services/FileDFG.jl [hunk 3]
```diff
@@ -123,14 +16,14 @@ function saveDFG(dfg::G, folder::String) where G <: AbstractDFG
...
-        vPacked = _packVariable(dfg, v)
+        vPacked = packVariable(dfg, v)
         io = open("$varFolder/$(v.label).json", "w")
         JSON2.write(io, vPacked)
         close(io)
     end
     # Factors
     for f in factors
-        fPacked = _packFactor(dfg, f)
+        fPacked = packFactor(dfg, f)
```
Comment (dehann) [inline src/FileDFG/services/FileDFG.jl]: very cool!

File: src/DFGPlots/DFGPlots.jl [hunk 4]
```diff
@@ -46,8 +67,27 @@ function dfgplot(dfg::LightDFG, p::DFGPlotProps = DFGPlotProps())
...
+# Using GraphViz plotting
+dfgplot(fg)
+# Save to PDFG
+using Compose
+draw(PDF("/tmp/graph.pdf", 16cm, 16cm), dfgplot(fg))
+```
+
+More information at [GraphPlot.jl](https://github.com/JuliaGraphs/GraphPlot.jl)
+"""
+function dfgplot(dfg::DistributedFactorGraphs.MetaGraphsDFG, p::DFGPlotProps = DFGPlotProps())
```
Comment (dehann) [inline src/DFGPlots/DFGPlots.jl]: excited to try this!

File: src/entities/DFGVariable.jl [hunk 5]
```diff
@@ -10,13 +14,15 @@ mutable struct VariableNodeData
   eliminated::Bool
   BayesNetVertID::Symbol #  Union{Nothing, }
   separator::Array{Symbol,1}
-  softtype
+  softtype::InferenceVariable #TODO v0.5.0 T
```
Comment (dehann) [inline src/entities/DFGVariable.jl]: yes, type T here is right way to go.

File: src/entities/DFGVariable.jl [hunk 6]
```diff
@@ -10,13 +14,15 @@ mutable struct VariableNodeData
...
   initialized::Bool
   inferdim::Float64
   ismargin::Bool
   dontmargin::Bool
+  # Tonio surprise TODO
+  # frontalonly::Bool
   # A valid, packable default constructor is needed.
-  VariableNodeData() = new(zeros(1,1), zeros(1,1), Symbol[], Int[], 0, false, :NOTHING, Symbol[], "", false, false, false, false)
+  VariableNodeData() = new(zeros(1,1), zeros(1,1), Symbol[], Int[], 0, false, :NOTHING, Symbol[], SingletonInferenceVariable(), false, false, false, false)
```
Comment (dehann) [inline src/entities/DFGVariable.jl]: The use of `Singleton` here is not that great, but can fix after v0.4.1.  `Singleton` relates to Factors not Variables

File: src/entities/DFGVariable.jl [hunk 7]
```diff
@@ -104,11 +110,47 @@ timestamp(v::DFGVariable) = v.timestamp
...
 internalId(v::DFGVariable) = v._internalId
 # Todo: Complete this.
 smallData(v::DFGVariable) = v.smallData
 bigData(v::DFGVariable) = v.bigData
+
+"""
+    $(SIGNATURES)
+Structure for first-class citizens of a DFGVariable.
+"""
+mutable struct DFGVariableSummary <: AbstractDFGVariable
```
Comment (dehann) [inline src/entities/DFGVariable.jl]: we're missing the `softtypename::Symbol` here -- is that in a future PR?  @Affie did you have strong preference here for visualization.  We did talk about this a bit.

File: src/services/AbstractDFG.jl [hunk 8]
```diff
@@ -61,79 +61,79 @@ end
     $(SIGNATURES)
 Add a DFGVariable to a DFG.
 """
-function addVariable!(dfg::G, variable::DFGVariable)::Bool where G <: AbstractDFG
+function addVariable!(dfg::G, variable::V)::Bool where {G <: AbstractDFG, V <: AbstractDFGVariable}
```
Comment (dehann) [inline src/services/AbstractDFG.jl]: Johan correctly pointed out that we should not overdo static parameters:
https://docs.julialang.org/en/v1/manual/style-guide/#Don't-use-unnecessary-static-parameters-1

Comment (dehann) [reply src/services/AbstractDFG.jl]: not serious, but keep in mind for future work.  I'm pretty guilty of this myself.

File: src/entities/DFGVariable.jl (hunk 6 above)
Comment (dehann) [reply src/entities/DFGVariable.jl]: For future reference, for single dimension variable or default, I use `ContinuousScalar`.

Comment (Affie) [reply src/entities/DFGVariable.jl]: I just named it Singleton from: Immutable composite types with no fields are singletons; 
I’m open to any suggestions. Default, empty.. etc. 
It should ideally only be temporary In use until the softtype becomes type parameter. 
Do you propose we move or duplicate ContinuousScalar to DFG?

**Pull request merged**